crates/iroha_multisig_data_model/src/lib.rs

```diff
@@ -36,39 +51,110 @@ type Weight = u8;
+        impl Instruction for $box {}
 
-    fn try_from(payload: &Json) -> serde_json::Result<Self> {
-        serde_json::from_str::<Self>(payload.as_ref())
-    }
+        impl From<$box> for InstructionBox {
+            fn from(value: $box) -> Self {
+                Self::Custom(value.into())
+            }
+        }
+
+        impl From<$box> for CustomInstruction {
+            fn from(value: $box) -> Self {
+                let payload = serde_json::to_value(&value)
+                    .expect(concat!("INTERNAL BUG: Couldn't serialize ", stringify!($box)));
+
+                Self::new(payload)
+            }
+        }
+
+        impl TryFrom<&Json> for $box {
+            type Error = serde_json::Error;
+
+            fn try_from(payload: &Json) -> serde_json::Result<Self> {
+                serde_json::from_str::<Self>(payload.as_ref())
+            }
+        }
+
+        $(
+            impl Instruction for $instruction {}
+
+            impl From<$instruction> for InstructionBox {
+                fn from(value: $instruction) -> Self {
+                    Self::Custom(<$box>::from(value).into())
+                }
+            }
+        )+
+    };
 }
 
-impl From<MultisigTransactionArgs> for Json {
-    fn from(details: MultisigTransactionArgs) -> Self {
-        Json::new(details)
+impl_custom_instruction!(
+    MultisigInstructionBox,
+    MultisigRegister | MultisigPropose | MultisigApprove
+);
+
+#[allow(missing_docs)]
+mod constants {
+    use super::*;
+
+    pub const SIGNATORIES: &str = "signatories";
+    pub const QUORUM: &str = "quorum";
+    pub const TRANSACTION_TTL_MS: &str = "transaction_ttl_ms";
+    pub const PROPOSALS: &str = "proposals";
+    pub const MULTISIG_SIGNATORY_: &str = "MULTISIG_SIGNATORY_";
+
+    pub fn instructions_key(hash: &HashOf<Vec<InstructionBox>>) -> Name {
+        format!("{PROPOSALS}/{hash}/instructions").parse().unwrap()
+    }
+
+    pub fn proposed_at_ms_key(hash: &HashOf<Vec<InstructionBox>>) -> Name {
+        format!("{PROPOSALS}/{hash}/proposed_at_ms")
+            .parse()
+            .unwrap()
     }
-}
 
-impl TryFrom<&Json> for MultisigTransactionArgs {
-    type Error = serde_json::Error;
+    pub fn approvals_key(hash: &HashOf<Vec<InstructionBox>>) -> Name {
+        format!("{PROPOSALS}/{hash}/approvals").parse().unwrap()
+    }
 
-    fn try_from(payload: &Json) -> serde_json::Result<Self> {
-        serde_json::from_str::<Self>(payload.as_ref())
+    pub fn multisig_role_for(account: &AccountId) -> RoleId {
+        format!(
+            "{MULTISIG_SIGNATORY_}{}_{}",
+            account.signatory(),
+            account.domain()
+        )
+        .parse()
+        .unwrap()
+    }
+
+    pub fn multisig_account_from(role: &RoleId) -> Option<AccountId> {
+        role.name()
+            .as_ref()
+            .strip_prefix(MULTISIG_SIGNATORY_)?
+            .replacen('_', "@", 1)
+            .parse()
+            .ok()
     }
 }
+
+pub use constants::*;
```

Comment:
as usual, I oppose making this public because this is not useful for consumers of this crate. This is just for devs. I'd be willing to let it slide if you mark everything with `#[doc(hidden)]` and use `__` prefix. But I'd rather look into removing it

Reply:
also, your use of these in `iroha` tests is an anti-pattern. IMO black box tests should test what an end-user might write themselves. They also serve the purpose of examples of how to use `iroha`

Reply:
Defined them as private in `iroha_default_executor::multisig`, `iroha_executor::default::role` and `iroha_cli::main` respectively